File: seminar_1/task_2/Ishmuratov.Rail/Task_2.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* reverseList(ListNode* head) {
    ListNode* prev = NULL;
    ListNode* curr = head;
    while(curr != NULL){
        ListNode* tmp = curr->next;
        curr->next = prev;
        prev = curr;
        curr = tmp;
    }
    return prev;
}

int main()
{
    ListNode d{4};
    ListNode c{3, &d};
    ListNode b{2, &c};
    ListNode a{1, &b}; // Это head
    ListNode* head = &a;
    while (head != nullptr) {
        cout<<head->val<<" ";
        head = head->next;
    }
    cout<<"\n";
    ListNode* result = reverseList(&a);
    while (result != nullptr) {
        cout<<result->val<<" ";
        result = result->next;
    }
    cout<<"\n";
}

File: seminar_1/task_2/patokova.victoria/task-2.py
class ListNode:
    def __init__(self, data):
        self.data = data
        self.next = None


def reverseLinkedList(head):
    prev = None
    cur = head
    while cur != None:
        temp = cur.next
        cur.next = prev
        prev = cur
        cur = temp

    head = prev
    return head


def printLinkedList(head):
    cur = head
    while cur != None:
        print(cur.data)
        cur = cur.next

        
e1 = ListNode(1)
e2 = ListNode(2)
e3 = ListNode(3)
e4 = ListNode(4)
e5 = ListNode(3)

e1.next = e2
e2.next = e3
e3.next = e4
e4.next = e5

head = e1


head = reverseLinkedList(head)
printLinkedList(head)